diff --git a/ckan/public/base/javascript/notify.js b/ckan/public/base/javascript/notify.js
index 1837a0a..901f8f4 100644
--- a/ckan/public/base/javascript/notify.js
+++ b/ckan/public/base/javascript/notify.js
@@ -36,7 +36,7 @@
    * Returns the element.
    */
   notify.create = function (title, message, type) {
-    var alert = jQuery('<div class="alert fade in"><strong></strong> <span></span></div>');
+    var alert = jQuery('<div class="alert alert-dismissable fade in"><strong></strong> <span></span></div>');
     alert.addClass('alert-' + (type || 'error'));
     alert.find('strong').text(title);
     alert.find('span').text(message);
@@ -51,7 +51,8 @@
    */
   notify.initialize = function (element) {
     element = element instanceof jQuery ? element : jQuery(element);
-    return element.append(jQuery('<a class="close" href="#">&times;</a>')).alert();
+    return element.append(jQuery('<button type="button" class="close" data-dismiss="alert" aria-label="Close"><span aria-hidden="true">&times;</span></button>'));
+
   };
 
   // Initialize any alerts already on the page.
