diff --git a/ckan/logic/auth/create.py b/ckan/logic/auth/create.py
index 31f67c3..f62bfdc 100644
--- a/ckan/logic/auth/create.py
+++ b/ckan/logic/auth/create.py
@@ -157,6 +157,8 @@ def user_create(context, data_dict=None):
     create_user_via_web = authz.check_config_permission(
             'create_user_via_web')
 
+    if 'group' in context:
+        return member_create(context, data_dict)
     if using_api and not create_user_via_api:
         return {'success': False, 'msg': _('User {user} not authorized to '
             'create users via the API').format(user=context.get('user'))}
