diff --git a/ckan/logic/converters.py b/ckan/logic/converters.py
index c87cad49a..a6d99d2a0 100644
--- a/ckan/logic/converters.py
+++ b/ckan/logic/converters.py
@@ -25,14 +25,24 @@ def convert_to_extras(key, data, errors, context):
 
 def convert_from_extras(key, data, errors, context):
 
-    def remove_from_extras(data, key):
-        to_remove = []
-        for data_key, data_value in data.iteritems():
-            if (data_key[0] == 'extras'
-                and data_key[1] == key):
-                to_remove.append(data_key)
-        for item in to_remove:
-            del data[item]
+    def remove_from_extras(data, idx):
+        for key in sorted(data):
+            if key[0] != 'extras' or key[1] < idx:
+                continue
+            if key[1] == idx:
+                del data[key]
+
+            # Following block required for unflattening extras with
+            # "gaps" created sometimes by `convert_from_extra`
+            # validator :
+            #
+            #   {
+            #     ('extras', 0, 'key'): 'x',
+            #     ('extras', 2, 'key): 'y'
+            #   }
+            if key[1] > idx:
+                new_key = (key[0], key[1] - 1) + key[2:]
+                data[new_key] = data.pop(key)
 
     for data_key, data_value in data.iteritems():
         if (data_key[0] == 'extras'
