diff --git a/ckan/lib/dictization/model_dictize.py b/ckan/lib/dictization/model_dictize.py
index 4211bc435..f6a5d0e47 100644
--- a/ckan/lib/dictization/model_dictize.py
+++ b/ckan/lib/dictization/model_dictize.py
@@ -200,7 +200,9 @@ def package_dictize(
         pkg_tag, tag.c["id"] == pkg_tag.c["tag_id"]
     ).where(pkg_tag.c["package_id"] == pkg.id)
     result = execute(q, pkg_tag, context)
-    result_dict["tags"] = d.obj_list_dictize(result, context,
+    dictize_context = context.copy()
+    dictize_context['with_capacity'] = False
+    result_dict["tags"] = d.obj_list_dictize(result, dictize_context,
                                              lambda x: x["name"])
     result_dict['num_tags'] = len(result_dict.get('tags', []))
 
diff --git a/ckan/logic/action/get.py b/ckan/logic/action/get.py
index 2f6effb4c..5f564c10f 100644
--- a/ckan/logic/action/get.py
+++ b/ckan/logic/action/get.py
@@ -617,7 +617,9 @@ def group_list_authz(context: Context,
         if package:
             groups = set(groups) - set(package.get_groups())
 
-    group_list = model_dictize.group_list_dictize(groups, context,
+    dictize_context = context.copy()
+    dictize_context['with_capacity'] = False
+    group_list = model_dictize.group_list_dictize(groups, dictize_context,
         with_package_counts=asbool(data_dict.get('include_dataset_count')),
         with_member_counts=asbool(data_dict.get('include_member_count')))
     return group_list
