diff --git a/ckan/lib/mailer.py b/ckan/lib/mailer.py
index 9eaf2fd..c75e343 100644
--- a/ckan/lib/mailer.py
+++ b/ckan/lib/mailer.py
@@ -144,11 +144,11 @@ def get_invite_body(user, group_dict=None, role=None):
 
 
 def get_reset_link(user):
-    return urljoin(config.get('site_url'),
-                   h.url_for(controller='user',
-                             action='perform_reset',
-                             id=user.id,
-                             key=user.reset_key))
+    return h.url_for(controller='user',
+                     action='perform_reset',
+                     id=user.id,
+                     key=user.reset_key,
+                     qualified=True)
 
 
 def send_reset_link(user):
