diff --git a/ckan/controllers/package.py b/ckan/controllers/package.py
index 473eba5..d256b96 100644
--- a/ckan/controllers/package.py
+++ b/ckan/controllers/package.py
@@ -336,7 +336,7 @@ class PackageController(base.BaseController):
     def read(self, id):
         context = {'model': model, 'session': model.Session,
                    'user': c.user or c.author, 'for_view': True,
-                   'auth_user_obj': c.userobj}
+                   'auth_user_obj': c.userobj, 'validate': False}
         data_dict = {'id': id, 'include_tracking': True}
 
         # interpret @<revision_id> or @<date> suffix
@@ -1077,7 +1077,8 @@ class PackageController(base.BaseController):
         context = {'model': model, 'session': model.Session,
                    'user': c.user or c.author,
                    'auth_user_obj': c.userobj,
-                   'for_view': True}
+                   'for_view': True,
+                   'validate': False}
 
         try:
             c.package = get_action('package_show')(context, {'id': id})
