diff --git a/ckan/lib/helpers.py b/ckan/lib/helpers.py
index d95b87c..233e89f 100644
--- a/ckan/lib/helpers.py
+++ b/ckan/lib/helpers.py
@@ -1045,11 +1045,7 @@ def dict_list_reduce(list_, key, unique=True):
 
 @core_helper
 def linked_gravatar(email_hash, size=100, default=None):
-    return literal(
-        '<a href="https://gravatar.com/" target="_blank" ' +
-        'title="%s" alt="">' % _('Update your avatar at gravatar.com') +
-        '%s</a>' % gravatar(email_hash, size, default)
-    )
+    return ''
 
 
 _VALID_GRAVATAR_DEFAULTS = ['404', 'mm', 'identicon', 'monsterid',
@@ -1058,17 +1054,7 @@ _VALID_GRAVATAR_DEFAULTS = ['404', 'mm', 'identicon', 'monsterid',
 
 @core_helper
 def gravatar(email_hash, size=100, default=None):
-    if default is None:
-        default = config.get('ckan.gravatar_default', 'identicon')
-
-    if default not in _VALID_GRAVATAR_DEFAULTS:
-        # treat the default as a url
-        default = urllib.quote(default, safe='')
-
-    return literal('''<img src="//gravatar.com/avatar/%s?s=%d&amp;d=%s"
-        class="gravatar" width="%s" height="%s" alt="" />'''
-                   % (email_hash, size, default, size, size)
-                   )
+    return ''
 
 
 @core_helper
