diff --git a/ckan/authz.py b/ckan/authz.py
index 270125512..09eb9b227 100644
--- a/ckan/authz.py
+++ b/ckan/authz.py
@@ -114,8 +114,12 @@ class AuthFunctions:
                 else:
                     # fallback to chaining off the builtin auth function
                     prev_func = self._functions[name]
-                fetched_auth_functions[name] = (
-                    functools.partial(func, prev_func))
+
+                wrapped_func = functools.partial(func, prev_func)
+                if getattr(func, 'auth_allow_anonymous_access', False):
+                    wrapped_func.auth_allow_anonymous_access = True
+
+                fetched_auth_functions[name] = wrapped_func
 
         # Use the updated ones in preference to the originals.
         self._functions.update(fetched_auth_functions)
