diff --git a/ckan/views/resource.py b/ckan/views/resource.py
index 7f22152d6..62e5574f1 100644
--- a/ckan/views/resource.py
+++ b/ckan/views/resource.py
@@ -171,7 +171,7 @@ def download(package_type: str,
     if rsc.get(u'url_type') == u'upload':
         upload = uploader.get_resource_uploader(rsc)
         filepath = upload.get_path(rsc[u'id'])
-        resp = flask.send_file(filepath, download_name=filename)
+        resp = flask.send_file(filepath, download_name=filename, as_attachment=True)

         if rsc.get('mimetype'):
             resp.headers['Content-Type'] = rsc['mimetype']
